batch/batch.go

```diff
@@ -0,0 +1,282 @@
+	}
+	return from, to
+}
+
+// Short gives a brief summary of the contained requests and keys in the batch.
+// TODO(tschottdorf): awkward to not have this on BatchRequest, but can't pull
+// `keys` into `proto` (req'd by KeyAddress). See #2198.
+func Short(br *proto.BatchRequest) string {
+	var str []string
+	for _, arg := range br.Requests {
+		req := arg.GetValue().(proto.Request)
+		h := req.Header()
+		str = append(str, fmt.Sprintf("%T [%s,%s)", req, h.Key, h.EndKey))
+	}
+	from, to := KeyRange(br)
+	return fmt.Sprintf("[%s,%s): ", from, to) + strings.Join(str, ", ")
+}
+
+// Sender is a new incarnation of client.Sender which only supports batches
+// and uses a request-response pattern.
+// TODO(tschottdorf): do away with client.Sender.
+// TODO(tschottdorf) rename to SendBatch->Send when client.Sender is out of the way.
+type Sender interface {
+	SendBatch(context.Context, *proto.BatchRequest) (*proto.BatchResponse, error)
+}
+
+// SenderFn is a function that implements a Sender.
+type SenderFn func(context.Context, *proto.BatchRequest) (*proto.BatchResponse, error)
+
+// SendBatch implements batch.Sender.
+func (f SenderFn) SendBatch(ctx context.Context, ba *proto.BatchRequest) (*proto.BatchResponse, error) {
+	return f(ctx, ba)
+}
+
+// A ChunkingSender sends batches, subdividing them appropriately.
+// TODO(tschottdorf): doesn't actually chunk much yet, only puts EndTransaction
+// into an extra batch. Easy to complete.
+// TODO(tschottdorf): only used by DistSender, but let's be modular.
+type ChunkingSender struct {
+	f SenderFn
+}
+
+// NewChunkingSender returns a new chunking sender which sends through the supplied
+// SenderFn.
+func NewChunkingSender(f SenderFn) Sender {
+	return &ChunkingSender{f: f}
+}
+
+// SendBatch implements Sender.
+// TODO(tschottdorf): only cuts an EndTransaction request off. Also need
+// to untangle reverse/forward, txn/non-txn, ...
+// We actually don't want to chop EndTransaction off for single-range
+// requests. Whether it is one or not is unknown right now (you can only
+// find out after you've sent to the Range/looked up a descriptor that
+// suggests that you're multi-range. In those cases, the wrapped sender should
+// return an error so that we split and retry once the chunk which contains
+// EndTransaction (i.e. the last one).
+func (cs *ChunkingSender) SendBatch(ctx context.Context, ba *proto.BatchRequest) (*proto.BatchResponse, error) {
+	var argChunks []*proto.BatchRequest
+	if len(ba.Requests) < 1 {
+		panic("empty batchArgs")
+	}
+
+	// Deterministically create ClientCmdIDs for all parts of the batch if
+	// a CmdID is already set (otherwise, leave them empty).
+	var nextID func() proto.ClientCmdID
+	empty := proto.ClientCmdID{}
+	if empty == ba.CmdID {
+		nextID = func() proto.ClientCmdID {
+			return empty
+		}
+	} else {
+		rng := rand.New(rand.NewSource(ba.CmdID.Random))
+		origID := ba.CmdID
+		nextID = func() proto.ClientCmdID {
+			id := origID
+			origID.Random = rng.Int63()
```

Comment:
it's intentional (I want to start with the original ID, updating that for the next call). Fixed up the naming (`origID`->`id`, `id`->`curID`).

I don't think we rely on entropy for the `Random` component, but this seemed safer.